Add ccoctl command to create workload identity pool in GCP

akhil-rane opened the pull request:
- [x] vendor google.golang.org/api/iam/v1, this has sdk support for workload identity
- [x] Fix tests in pkg/operator/utils/gcp/utils_test.go due gomock version update
- [x] Move common code in pkg/cmd/provisioning package to utils.go
- [x] Add ccoctl gcp command to create workload identity pool

x-ref: https://issues.redhat.com/browse/CCO-119

akhil-rane commented:
/hold

wait for #367

akhil-rane commented:
/test e2e-gcp

joelddiaz commented on pkg/operator/utils/gcp/utils_test.go:
```diff
@@ -139,7 +139,7 @@ func TestCheckPermissionsAgainstPermissionListChunking(t *testing.T) {
 				chunkCopy := chunk
 
-				mockCall := mockClient.EXPECT().TestIamPermissions("fake", gomock.Any()).DoAndReturn(
+				mockCall := mockClient.EXPECT().TestIamPermissions("fake", gomock.Any()).AnyTimes().DoAndReturn(
```
Should these really be `AnyTimes()` or could they be `Times(1)` (just to keep the test case expectations tighter)?

akhil-rane replied:
This was helpful. It caught some failing tests. This might need more careful review.

joelddiaz commented on pkg/gcp/client.go:
```diff
@@ -149,6 +152,10 @@ func (c *gcpClient) QueryTestablePermissions(ctx context.Context, request *iamad
 }
 
+func (c *gcpClient) CreateWorkloadIdentityPool(ctx context.Context, parent, poolID string, pool *iam.WorkloadIdentityPool) (*iam.Operation, error) {
+	return c.iamService.Projects.Locations.WorkloadIdentityPools.Create(parent, pool).WorkloadIdentityPoolId(poolID).Context(ctx).Do()
```
lets wrap this with a timeout like the other wrapper functions do
```suggestion
	ctx, cancel := contextWithTimeout(ctx)
	defer cancel()
	return c.iamService.Projects.Locations.WorkloadIdentityPools.Create(parent, pool).WorkloadIdentityPoolId(poolID).Context(ctx).Do()
```

joelddiaz commented on pkg/cmd/provisioning/gcp/create_workload_identity_pool.go:
```diff
@@ -0,0 +1,75 @@
+		Run:   createWorkloadIdentityPoolCmd,
+	}
+
+	createIAMRolesCmd.PersistentFlags().StringVar(&CreateWorkloadIdentityPoolOpts.Name, "name", "", "User-define name for all created Google cloud resources (can be separate from the cluster's infra-id)")
```
```suggestion
	createIAMRolesCmd.PersistentFlags().StringVar(&CreateWorkloadIdentityPoolOpts.Name, "name", "", "User-defined name for all created Google cloud resources (can be separate from the cluster's infra-id)")
```

joelddiaz commented on pkg/cmd/provisioning/gcp/create_workload_identity_pool.go:
```diff
@@ -0,0 +1,75 @@
+
+	gcpClient, err := gcp.NewClient(CreateWorkloadIdentityPoolOpts.Project, nil)
+	if err != nil {
+		log.Fatal(err)
```
```suggestion
		log.Fatalf("Failed to setup GCP client: %s", err)
```

joelddiaz commented on pkg/cmd/provisioning/gcp/create_workload_identity_pool.go:
```diff
@@ -0,0 +1,75 @@
+		Disabled:    false,
+	}
+
+	_, err := client.CreateWorkloadIdentityPool(ctx, fmt.Sprintf("projects/%s/locations/global", project), shortenedName, pool)
```
can we make a simple string template to store the magic "projects/%s/locations/global" string (with a hopefully descriptive variable name)?

joelddiaz commented on pkg/cmd/provisioning/gcp/create_workload_identity_pool_test.go:
```diff
@@ -0,0 +1,79 @@
+	mockGCPClient.EXPECT().CreateWorkloadIdentityPool(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(
+		&iam.Operation{
+			Done: true,
+		}, nil).AnyTimes()
```
Can we tighten this expect to be more strict than `AnyTimes()`?

joelddiaz commented on pkg/cmd/provisioning/gcp/gcp.go:
```diff
@@ -0,0 +1,31 @@
+// NewGCPCmd implements the "gcloud" subcommand for the credentials provisioning
+func NewGCPCmd() *cobra.Command {
+	createCmd := &cobra.Command{
+		Use:   "gcloud",
```
I don't have a strong opinion here, but I expected it to be `gcp` not `gcloud`. What do others think?

dgoodwin replied:
+1 for gcp

akhil-rane replied:
gcp it is

abhinavdahiya commented on pkg/gcp/client.go:
```diff
@@ -198,9 +205,16 @@ func fixupServiceMap(serviceMap map[string]bool) {
-	creds, err := google.CredentialsFromJSON(context.TODO(), authJSON, "https://www.googleapis.com/auth/cloud-platform")
+	scope := "https://www.googleapis.com/auth/cloud-platform"
+
+	if authJSON != nil {
```
check for len of slices instead of nil value.

akhil-rane commented:
/test e2e-gcp

akhil-rane commented on pkg/operator/utils/gcp/utils_test.go:
```diff
@@ -171,39 +182,20 @@ func TestIgnoreInvalidProjectPermissions(t *testing.T) {
-				// now allow the perm list w/o invalidPerm
-				secondMock := mockClient.EXPECT().TestIamPermissions(projectName, gomock.Any()).DoAndReturn(
+				// allow the perm list w/o invalidPerm
+				mockCall := mockClient.EXPECT().TestIamPermissions(projectName, gomock.Any()).Times(1).DoAndReturn(
```
There is only one call to `TestIamPermissions` inside `CheckPermissionsAgainstPermissionList`. It takes all permissions and returns valid permissions. I do not think other mocks are required here.

joelddiaz replied:
Is this being missed because of a missing `defer Finish()` on the generated mockcontroller down by line 255?

akhil-rane replied:
I am not really sure why we are expecting 2 calls to `TestIamPermissions` here. From what I understand (from the code snippet below of `CheckPermissionsAgainstPermissionList` function), we should have 1 call to `TestIamPermissions` for 100 permissions (i.e. chunksize) and since we have only 2 permissions there should be only 1 call. 

```
	chunkSize := 100
	permLen := len(filteredPermList)
	for i := 0; i < permLen; i += chunkSize {
		end := i + chunkSize
		if end > permLen {
			end = permLen
		}
		req := &cloudresourcemanager.TestIamPermissionsRequest{Permissions: filteredPermList[i:end]}
		resp, err := gcpClient.TestIamPermissions(projectName, req)
		if err != nil {
			return false, fmt.Errorf("error testing permissions: %v", err)
		}
		allowedPerms.Insert(resp.Permissions...)
	}
```

Correct me if I am missing something here :(

joelddiaz replied:
Maybe I'm missing something in this discussion. I only see one EXPECT() for TestIamPermissions()...

joelddiaz replied:
Ok, we got this figured out after talking.

joelddiaz commented on pkg/cmd/provisioning/gcp/create_workload_identity_pool.go:
```diff
@@ -0,0 +1,118 @@
+		createWorkloadIdentityPoolScript := provisioning.CreateShellScript([]string{
+			fmt.Sprintf(createIdentityPoolCmd, name, name),
+		})
+		createIdentityPoolScripFullPath := filepath.Join(targetDir, provisioning.CreateIdentityPoolScriptName)
+		if err := provisioning.SaveToFile(fmt.Sprintf("script to create workload identity pool %s", name), createIdentityPoolScripFullPath, []byte(createWorkloadIdentityPoolScript)); err != nil {
```
```suggestion
		createIdentityPoolScriptFullPath := filepath.Join(targetDir, provisioning.CreateIdentityPoolScriptName)
		if err := provisioning.SaveToFile(fmt.Sprintf("script to create workload identity pool %s", name), createIdentityPoolScriptFullPath, []byte(createWorkloadIdentityPoolScript)); err != nil {
```

joelddiaz commented on pkg/cmd/provisioning/gcp/create_workload_identity_pool_test.go:
```diff
@@ -0,0 +1,130 @@
+
+import (
+	"context"
+	"github.com/openshift/cloud-credential-operator/pkg/cmd/provisioning"
```
move this import down with the mockgcp import

joelddiaz commented on pkg/cmd/provisioning/utils.go:
```diff
@@ -23,3 +27,34 @@ func EnsureDir(path string) error {
+}
+
+// CountNonDirectoryFiles counts files which are not a directory
+func CountNonDirectoryFiles(files []os.FileInfo) int {
```
Is this used outside of the test files?

joelddiaz replied:
that's fair. let each test file at least share this functionality :+1:

akhil-rane replied:
fixed

joelddiaz commented on pkg/cmd/provisioning/utils.go:
```diff
@@ -23,3 +27,34 @@ func EnsureDir(path string) error {
+}
+
+// SaveToFile saves the given data to a given file
+func SaveToFile(filePurpose, filePath string, data []byte) error {
```
Any particular reason to not simply use ioutil.WriteFIle()?

akhil-rane replied:
to be honest I just moved it to utils.go since it was a common code. Let me explore if we can use ioutil.WriteFile() here.

akhil-rane replied:
done

joelddiaz commented on pkg/operator/utils/gcp/utils_test.go:
```diff
@@ -107,6 +107,14 @@ func perms(low, hi int) []string {
 }
 
+func createStringSet(stringArray []string) sets.String {
+	stringSet := sets.NewString()
+	for _, str := range stringArray {
+		stringSet.Insert(str)
+	}
+	return stringSet
```
I think you can pass in your initial list of strings into the NewString() call.
```suggestion
        return sets.NewString(stringArray...)
```

joelddiaz reviewed (commented):
One overarching comment on saving the gcloud commands to a script is that this really makes our approach a Mac/Linux solution (unless we consider WSL acceptable for Windows users). 
We don't presently produce ccoctl for anything except Linux today (due to where in the build pipeline this repo is processed), but it has been discussed previously to build ccoctl for the same target set as the installer is built for.

Thoughts?

akhil-rane commented on cmd/ccoctl/main.go:
```diff
@@ -19,6 +19,7 @@ func main() {
 	rootCmd.AddCommand(aws.NewAWSCmd())
 	rootCmd.AddCommand(gcp.NewGCPCmd())
 	rootCmd.AddCommand(ibmcloud.NewIBMCloudCmd())
+	rootCmd.AddCommand(gcp.NewGCPCmd())
```
yes, will fix it

akhil-rane commented:
@joelddiaz this is ready for another round of review

/assign @jeana-redhat @lwan-wanglin @sferich888

jeana-redhat commented:
/label docs-approved

akhil-rane commented on pkg/cmd/provisioning/constants.go:
```diff
@@ -13,4 +13,6 @@ const (
 	// TLSDirName is the name of the directory to save bound service account signing key created by ccoctl
 	TLSDirName = "tls"
+	// FileModeCcoctlDryRun represents a mode and permission bits of the files created by ccoctl in dry run
+	FileModeCcoctlDryRun = 0744
```
you are right! I will keep separate constants for aws and gcp

akhil-rane replied:
@joelddiaz this is fixed.

lwan-wanglin commented:
@akhil-rane  here is the test case for the PR https://url.corp.redhat.com/9aab015, could you help to review?

akhil-rane commented:
> @akhil-rane here is the test case for the PR https://url.corp.redhat.com/9aab015, could you help to review?

@lwan-wanglin I think there should be one more step for `--dry-run`. After checking the contents of the shell script (step 6), we should run the shell script and check if workload identity pool is getting created in the gcp. wdyt?

joelddiaz reviewed (commented):
/lgtm

joelddiaz commented:
/assign @jeana-redhat @lwan-wanglin @sferich888 
As far as docs (and px), there is a standalone PR in flight with the documentation that should cover this whole series of PRs adding support for GCP workload-identity in #376

lwan-wanglin commented:
> I think there should be one more step for `--dry-run`. After checking the contents of the shell script (step 6), we should run the shell script and check if workload identity pool is getting created in the gcp. wdyt?

@akhil-rane thanks for your advice, I have updated the test case and added steps for running shell script, please check again https://url.corp.redhat.com/9aab015 .

And I have a small question,  openshift-install binary can read gcp credentials from /home/<user>/.gcp/osServiceAccount.json by default, is it possible for ccoctl to do so?  now ccoctl only respect GOOGLE_APPLICATION_CREDENTIALS env variable.

akhil-rane commented:
> > I think there should be one more step for `--dry-run`. After checking the contents of the shell script (step 6), we should run the shell script and check if workload identity pool is getting created in the gcp. wdyt?
> 
> @akhil-rane thanks for your advice, I have updated the test case and added steps for running shell script, please check again https://url.corp.redhat.com/9aab015 .
> 
> And I have a small question, openshift-install binary can read gcp credentials from /home//.gcp/osServiceAccount.json by default, is it possible for ccoctl to do so? now ccoctl only respect GOOGLE_APPLICATION_CREDENTIALS env variable.

interesting find about default location for credentials. I think we should include it. But it looks unrelated in this PR. Lets open a jira issue and handle it in a separate PR. Does that sound good to you?

akhil-rane commented:
/hold cancel

lwan-wanglin commented:
soulds ok to me.
/label qe-approved

sferich888 commented:
/label px-approved
